src/main/native/org_cryptomator_windows_keychain_WinHello_Native.cpp

```diff
@@ -0,0 +1,206 @@
+                ::SetForegroundWindow(hWnd);
+                g_promptFocusCount = 0;
+                break;
+            } else if (++g_promptFocusCount > 3) {
+                g_promptFocusCount = 0;
+                break;
+            }
+        }
+    }).detach();
+}
+
+void InitializeWindowsRuntime() {
+    std::call_once(runtimeInitFlag, []() {
+        winrt::init_apartment();
+    });
+}
+
+bool deriveEncryptionKey(const std::vector<uint8_t>& challenge, std::vector<uint8_t>& key){
+  auto challengeBuffer = CryptographicBuffer::CreateFromByteArray(
+        array_view<const uint8_t>(challenge.data(), challenge.size()));
+
+  try {
+    // The first time this is used a key-pair will be generated using the common name
+    auto result = KeyCredentialManager::RequestCreateAsync(s_winHelloKeyName,
+          KeyCredentialCreationOption::FailIfExists).get();
+
+    if (result.Status() == KeyCredentialStatus::CredentialAlreadyExists) {
+      result = KeyCredentialManager::OpenAsync(s_winHelloKeyName).get();
+    } else if (result.Status() != KeyCredentialStatus::Success) {
+      std::cout << "Failed to retrieve Windows Hello credential." << std::endl;
+      return false;
+    }
+
+    const auto signature = result.Credential().RequestSignAsync(challengeBuffer).get();
+    if (signature.Status() != KeyCredentialStatus::Success) {
+      if (signature.Status() != KeyCredentialStatus::UserCanceled) {
+        std::cout << "Failed to sign challenge using Windows Hello." << std::endl;
+      }
+      return false;
+    }
+
+    // Use the SHA-256 hash of the challenge signature as the encryption key
+    const auto response = signature.Result();
+    HashAlgorithmProvider hashProvider = HashAlgorithmProvider::OpenAlgorithm(HashAlgorithmNames::Sha256());
+    IBuffer hashBuffer = hashProvider.HashData(response);
+    key = iBufferToVector(hashBuffer);
+    return true;
+
+  } catch (winrt::hresult_error const& ex) {
+    std::cout << winrt::to_string(ex.message()) << std::endl;
+    return false;
+  }
+}
+
+// Encrypts data using Windows Hello KeyCredentialManager API and derived key from signed challenge
+jbyteArray JNICALL Java_org_cryptomator_windows_keychain_WinHello_00024Native_setEncryptionKey
+(JNIEnv* env, jobject obj, jbyteArray cleartext, jbyteArray challenge) {
+  queueSecurityPromptFocus();
+  try {
+    // Convert Java byte arrays to C++ vectors
+    std::vector<uint8_t> cleartextVec = jbyteArrayToVector(env, cleartext);
+    std::vector<uint8_t> challengeVec = jbyteArrayToVector(env, challenge);
+
+    InitializeWindowsRuntime();
+
+    // Take the random challenge and sign it by Windows Hello
+    // to create the key. The challenge is also used as the IV.
+    std::vector<uint8_t> key;
+    if (!deriveEncryptionKey(challengeVec, key)) {
+      throw std::runtime_error("Failed to generate the encryption key with the Windows Hello credential.");
+    }
+
+    auto algorithmName = SymmetricAlgorithmNames::AesCbcPkcs7();
```

Comment:
This is a quite old algorithm, use [AesGCM](https://learn.microsoft.com/en-us/uwp/api/windows.security.cryptography.core.symmetricalgorithmnames.aesgcm?view=winrt-26100#windows-security-cryptography-core-symmetricalgorithmnames-aesgcm)

Reply:
> This is a quite old algorithm, use [AesGCM](https://learn.microsoft.com/en-us/uwp/api/windows.security.cryptography.core.symmetricalgorithmnames.aesgcm?view=winrt-26100#windows-security-cryptography-core-symmetricalgorithmnames-aesgcm)

I am afraid, that does not work. It throws an error with `AesGcm`:
```
Error: Nicht implementiert (HRESULT: 0x80004001)
```

Reply:
Resolved with 9ac7cb65323f6758a60ed883c9653b9a2d1b366c.